src/Statement/StatementByGuidMap.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace Wikibase\DataModel\Statement;
+
+use ArrayIterator;
+use Countable;
+use InvalidArgumentException;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * Ordered and unique collection of Statement objects.
+ * Can only contain Statements that have a non-null GUID.
+ *
+ * Provides indexed access by Statement GUID.
+ * Does not provide complex modification functionality.
+ *
+ * @since 3.0
+ *
+ * @licence GNU GPL v2+
+ * @author Jeroen De Dauw < jeroendedauw@gmail.com >
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ */
+class StatementByGuidMap implements IteratorAggregate, Countable {
+
+	private $statements = array();
+
+	/**
+	 * @param Statement[] $statements
+	 */
+	public function __construct( $statements = array() ) {
+		foreach ( $statements as $statement ) {
+			$this->addStatement( $statement );
+		}
+	}
+
+	/**
+	 * If the provided statement has a GUID not yet in the map, it will be appended to the map.
+	 * If the GUID is already in the map, the statement with this guid will be replaced.
+	 *
+	 * @param Statement $statement
+	 */
+	public function addStatement( Statement $statement ) {
+		if ( $statement->getGuid() === null ) {
```

Comment:
This seems inconsistent. We allow GUIDs represented by e. g. integers here. If we try to get a statement from the list later, an exception is thrown, because it's not a string.